universe/Building.h

```diff
@@ -214,9 +214,8 @@ class BuildingTypeManager {
 
 private:
     BuildingTypeManager();
-    ~BuildingTypeManager();
 
-    std::map<std::string, BuildingType*> m_building_types;
+    std::map<std::string, std::unique_ptr<BuildingType>> m_building_types;
```

Comment:
The shared pointer can't be safely dereferenced.  It doesn't matter about the type of object.  I can't think of any exceptions, data members, function members, static functions and type info read/write could cause a crash.

Consider a simple case of `shared_ptr<int>`, with two different threads reading the `int`.  

Done correctly, with a mutex or some other technique thread safety, the compiler can guarantee at a hardware level, caches are coherent, reads are well ordered etc. and the two reads will not interfere with each other and will both return the same `int` value.  

Without some sort of thread safely mechanism dereferencing that pointer is undefined behavior, from a best case of two correct reads, to a worst case of a crash.